src/protocols/secure_channel/CASESessionCache.h

```diff
@@ -0,0 +1,42 @@
+/*
+ *
+ *    Copyright (c) 2021 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#pragma once
+
+#include <lib/core/CHIPError.h>
+#include <lib/core/PeerId.h>
+#include <protocols/secure_channel/CASESession.h>
+
+namespace chip {
+
+class CASESessionCache
+{
+public:
+    CASESessionCache();
+    virtual ~CASESessionCache();
+
+    CHIP_ERROR Add(CASESession & session);
```

Comment:
I'm considering the use case of this class.

When the secure session pool is exhausted, and we decide to hibernate a session, at the moment, where is the `CASESession` object ? I think the `CASESession` object is freed when the session is established.

So I'm not sure why we take a `CASESession` object here.

Reply:
I need to look into this and get back.

Reply:
Currently, the CASESession object is not being freed on session establishment. It's a member of OperationalDeviceProxy class.

If we decide to free it, it should be added to the cache prior to freeing it, as SecureSession object by itself doesn't carry all the information for caching and resumption.

Reply:
IMHO, we should free `CASESession` after the session is established, and copy required field into `SecureSession` object.

This is asking too much in this PR, but as least, I think this interface should take a `SecureSession` instead of a `CASESession`.

Reply:
SecureSession is not the correct class to pass to this interface. Resumption cache only works for sessions established via CASE. SecureSession doesn't have all the values needed for CASE resumption.